crates/bevy_pbr/src/light.rs

```diff
@@ -52,6 +53,84 @@ impl PointLightUniform {
     }
 }
 
+/// A Directional light.
+///
+/// Directional lights don't exist in reality but they are a good
+/// approximation for light sources VERY far away, like the sun or
+/// the moon.
+#[derive(Debug, Reflect)]
+#[reflect(Component)]
+pub struct DirectionalLight {
+    pub color: Color,
+    pub intensity: f32,
+    direction: Vec3,
+}
+
+impl DirectionalLight {
+    /// Create a new directional light component.
+    ///
+    /// # Panics
+    /// Will panic if `direction` is not normalized.
+    pub fn new(color: Color, intensity: f32, direction: Vec3) -> Self {
+        assert!((direction.length_squared() - 1.0).abs() < 0.0001);
+        DirectionalLight {
+            color,
+            intensity,
+            direction,
+        }
+    }
+
+    /// Set direction of light.
+    ///
+    /// # Panics
+    /// Will panic if `direction` is not normalized.
+    pub fn set_direction(&mut self, direction: Vec3) {
+        assert!((direction.length_squared() - 1.0).abs() < 0.0001);
+        self.direction = direction;
+    }
+
+    pub fn get_direction(&self) -> Vec3 {
+        self.direction
+    }
+}
+
+impl Default for DirectionalLight {
+    fn default() -> Self {
+        DirectionalLight {
+            color: Color::rgb(1.0, 1.0, 1.0),
+            intensity: 100000.0, // good start for a sun light
+            direction: Vec3::new(0.0, -1.0, 0.0),
+        }
+    }
+}
+
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct DirectionalLightUniform {
+    pub dir: [f32; 4],
+    pub color: [f32; 4],
+}
+
+unsafe impl Byteable for DirectionalLightUniform {}
+
+impl DirectionalLightUniform {
+    pub fn from(light: &DirectionalLight) -> DirectionalLightUniform {
```

Comment:
This should be it's own `impl` block for the `From` trait.

```rust
impl From<DirectionalLight> for DirectionalLightUniform {}
```

Reply:
I did it the same as PointLightUniform but that makes sense. If I implement From,  should I do it for PointLight too?

Reply:
Oh, actually, PointLightUniform's `from` as more than one parameter. Even if DirectionalLight needs only one, shouldn't it be better for the code of both lights to be symmetric? I think it would make reading much easier.

Reply:
Hmmm...
Tbh I think the `PointLightUniform::from` is wrong and should be `PointLightUniform::new`. (IMO `from` should be reserved for the `From` trait). 

However, in that case I guess this is ok :smile:

Reply:
Agreed. `from` should be for the From trait. If `from` requires extra parameters, it should be called new() (or from_x / with_x if it shouldn't be the "default" constructor)

Reply:
I renamed both to new. From expects T, not &T, so it wasn't valid for DirectionalLightUniform either.